posts/2022-01-03-kryo-pitfalls.adoc

```diff
@@ -0,0 +1,289 @@
+turned off and the class name is serialized inside the payload), which could also lead to problems.
+
+Imagine we have tree components: a Device, a Processor, and a Consumer. Each send information to the next
+using Kryo (let's ignore the medium for now).
+
+----
++----------------+      +----------------+      +----------------+
+|     Device     |----->|    Processor   |----->|    Consumer    |
++----------------+      +----------------+      +----------------+
+----
+
+And that the _Device_ declares the classes from the previous section: `Message`, `Signal`, etc. These are then
+aggregated by the _Processor_ which defines its own classes: `Event` and `Notification`. Which are then consumed
+by the _Consumer_ (this latter component doesn't care about the first set of types).
+
+When registering classes, we cannot reuse IDs, otherwise we'd run into conflict or disagreement. For example,
+imagine that both the _Device_ and the _Processor_ uses ID 10 for `Message` and that the _Processor_ use ID 11 for `Event`.
+But after a few months of running in production, the _Device_ decides to use ID 11, which breaks the _Processor_'s
+operation.
+
+This, together with the performance requirement of using https://github.com/EsotericSoftware/kryo#pooling[Pooling],
+led _us_ to try to stuff all registration inside a single pool, for which we created another common component
+to hold and register every definition, in hopes to avoid this problem. If every class registration is in a
+single place, there is no way writers and readers can disagree on the ID of a class.
+
+----
++----------------+      +----------------+      +----------------+
+|     Device     |----->|    Processor   |----->|    Consumer    |
++-------+--------+      +-------+--------+      +-------+--------+
+        |                       |                       |
+        |                       | (uses)                |
+        |                       |                       |
+        |               +----------------+              |
+        |    (uses)     |    Models &    |    (uses)    |
+        +-------------->|  Serialization |<-------------+
+                        +----------------+
+----
+
+Now the problem is that once the number of classes starts to grow this becomes harder to manage. Every change
+in any given class requires an update this common library, even if the downstream component doesn't use or
+care about the change.
+
+**Conclusion:** How do we solve this issue? We've been bouncing ideas for several months, none of which entirely satisfied us.
+Eventually I realised that this is a design problem, not an implementation one. We don't need to stuff
+everything into a single common component, as the boundaries between systems is _now_ clearly defined. A
+better solution would be for each component to expose both its models and serialization pool, since it's
+the component writing to the medium the one that knows which classes effectively end up there.
```

Comment:
Missing word?

```suggestion
everything into a single common component, as the boundaries between systems is _now_ clearly defined. A
better solution would be for each component to expose both its models and serialization pool, since it's
the component writing to the medium is the one that knows which classes effectively end up there.
```